synapse/appservice/scheduler.py

```diff
@@ -136,10 +138,17 @@ async def _send_request(self, service: ApplicationService):
         self.requests_in_flight.add(service.id)
         try:
             while True:
-                events = self.queued_events.pop(service.id, [])
-                ephemeral = self.queued_ephemeral.pop(service.id, [])
+                all_events = self.queued_events.get(service.id, [])
+                events = all_events[:MAX_EVENTS_PER_TRANSACTION]
+                del all_events[MAX_EVENTS_PER_TRANSACTION:]
```

Comment:
I think this should be:
```suggestion
                del all_events[:MAX_EVENTS_PER_TRANSACTION]
```
instead. Considering the example:

```python
    MAX_EVENTS_PER_TRANSACTION = 5
    ...
    all_events = self.queued_events.get(service.id, [])  # all_events is [0,1,2,3,4,5]
    events = all_events[:MAX_EVENTS_PER_TRANSACTION]  # event is [0,1,2,3,4] (we're sending these to the AS)
    del all_events[MAX_EVENTS_PER_TRANSACTION:]  # all_events becomes [0,1,2,3,4], we wanted [5].
```

likewise below.